File: shared/app-version.mjs
// App 版本与 profile 的唯一判定逻辑（document/design/desktop-app-packaging.md §4、§5）。
// 完整 SemVer 只有一个来源：package.json#version，由 scripts/sync-app-version.mjs
// 生成到 app-version.generated.mjs 和 src-tauri/src/app_version_generated.rs。
// 运行时不得从 Info.plist、文件名、Git tag、updater 响应或 package.json 现读版本。

// 本模块只放判定逻辑，不导入生成文件：生成器自己要用这些函数，反向依赖会让首次生成死锁。
// 需要常量的一方直接 import "./app-version.generated.mjs"。

export const PROFILE_PRODUCTION = "production";
export const PROFILE_BETA = "beta";
export const APP_PROFILES = [PROFILE_PRODUCTION, PROFILE_BETA];

// 官方 SemVer 正则，但不接受 build metadata：只差 `+xxx` 的两个版本对 updater
// 是同一个版本，允许它等于允许发布两个无法区分的产物。
const SEMVER_PATTERN =
  /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?$/;

export function parseAppVersion(version) {
  if (typeof version !== "string") {
    throw new Error("App version must be a string");
  }
  if (version.includes("+")) {
    throw new Error(
      `App version must not carry build metadata: ${version}；只差 build metadata 的版本对 updater 无法区分`,
    );
  }
  const match = SEMVER_PATTERN.exec(version);
  if (!match) {
    throw new Error(`App version must be a complete SemVer without a leading v, got: ${version}`);
  }
  const [, major, minor, patch, prerelease] = match;
  return {
    version,
    major: Number(major),
    minor: Number(minor),
    patch: Number(patch),
    prerelease: prerelease ?? null,
  };
}

// 构建时固定：无 pre-release 标记发 production，有则发 beta，运行时不可切换。
export function profileForVersion(version) {
  return parseAppVersion(version).prerelease === null ? PROFILE_PRODUCTION : PROFILE_BETA;
}

export function assertAppProfile(profile) {
  if (!APP_PROFILES.includes(profile)) {
    throw new Error(`App profile must be one of ${APP_PROFILES.join(", ")}, got: ${profile}`);
  }
  return profile;
}

// Tauri 与 sidecar 的启动握手：两端的完整版本字符串必须逐字符相等，
// 且该版本推导出的 profile 必须与传入的 profile 一致。
export function assertAppVersionHandshake({ appVersion, profile, expectedVersion }) {
  parseAppVersion(expectedVersion);
  if (appVersion !== expectedVersion) {
    throw new Error(
      `App version handshake failed: shell reported ${appVersion}, sidecar was built as ${expectedVersion}`,
    );
  }
  assertAppProfile(profile);
  const derived = profileForVersion(appVersion);
  if (derived !== profile) {
    throw new Error(
      `App version ${appVersion} belongs to the ${derived} profile, but ${profile} was requested`,
    );
  }
  return { appVersion, profile };
}

File: src-tauri/src/app_version.rs
// App 版本与 profile 推导（document/design/desktop-app-packaging.md §4、§5）。
// 唯一来源是 package.json#version，经 scripts/sync-app-version.mjs 生成到
// app_version_generated.rs。运行时不读 Info.plist 的 CFBundleShortVersionString：
// 打包工具可能把它规范化，pre-release 标记会丢，profile 判断随之失效。

include!("app_version_generated.rs");

pub const PROFILE_PRODUCTION: &str = "production";
pub const PROFILE_BETA: &str = "beta";

pub const PORT_PRODUCTION: u16 = 47824;
pub const PORT_BETA: u16 = 47825;

/// 构建时固定：无 pre-release 标记发 production，有则发 beta，运行时不可切换。
pub fn profile_for(version: &str) -> &'static str {
    if version.contains('-') {
        PROFILE_BETA
    } else {
        PROFILE_PRODUCTION
    }
}

pub fn profile() -> &'static str {
    profile_for(APP_VERSION_FULL)
}

pub fn port_for(profile: &str) -> u16 {
    if profile == PROFILE_BETA {
        PORT_BETA
    } else {
        PORT_PRODUCTION
    }
}

pub fn port() -> u16 {
    port_for(profile())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn stable_versions_use_the_production_profile() {
        assert_eq!(profile_for("0.1.0"), PROFILE_PRODUCTION);
        assert_eq!(profile_for("2.1.0"), PROFILE_PRODUCTION);
        assert_eq!(port_for(profile_for("2.1.0")), PORT_PRODUCTION);
    }

    #[test]
    fn pre_release_versions_use_the_beta_profile() {
        assert_eq!(profile_for("2.1.0-beta.1"), PROFILE_BETA);
        assert_eq!(profile_for("2.1.0-rc.1"), PROFILE_BETA);
        assert_eq!(port_for(profile_for("2.1.0-beta.1")), PORT_BETA);
    }

    // pre-release 编号必须逐字符保留，否则 beta.1 和 beta.2 会被当成同一个版本。
    #[test]
    fn pre_release_builds_stay_distinguishable() {
        assert_ne!("2.1.0-beta.1", "2.1.0-beta.2");
        assert_eq!(profile_for("2.1.0-beta.1"), profile_for("2.1.0-beta.2"));
    }

    #[test]
    fn the_generated_constant_is_a_usable_version() {
        assert!(!APP_VERSION_FULL.is_empty());
        assert!(!APP_VERSION_FULL.starts_with('v'));
        assert_eq!(APP_VERSION_FULL.split('.').count() >= 3, true);
    }
}
